File: nestack-backend/src/main/kotlin/com/nestack/domain/admin/service/AdminService.kt
package com.nestack.domain.admin.service

import com.nestack.common.constant.ErrorCode
import com.nestack.common.enum.AdminRole
import com.nestack.common.enum.AdminStatus
import com.nestack.common.enum.MissionStatus
import com.nestack.common.enum.UserStatus
import com.nestack.common.exception.BusinessException
import com.nestack.common.util.CryptoUtil
import com.nestack.domain.admin.dto.*
import com.nestack.domain.badge.service.BadgeService
import com.nestack.infrastructure.persistence.entity.*
import com.nestack.infrastructure.persistence.repository.*
import com.nestack.infrastructure.security.JwtTokenProvider
import org.slf4j.LoggerFactory
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.*

@Service
class AdminService(
    private val adminUserRepository: AdminUserRepository,
    private val userRepository: UserRepository,
    private val familyGroupRepository: FamilyGroupRepository,
    private val missionRepository: MissionRepository,
    private val categoryRepository: LifeCycleCategoryRepository,
    private val templateRepository: MissionTemplateRepository,
    private val badgeRepository: BadgeRepository,
    private val userBadgeRepository: UserBadgeRepository,
    private val announcementRepository: AnnouncementRepository,
    private val jwtTokenProvider: JwtTokenProvider,
    private val badgeService: BadgeService
) {
    private val logger = LoggerFactory.getLogger(AdminService::class.java)

    @Transactional
    fun login(request: AdminLoginRequest): AdminLoginResponse {
        val admin = adminUserRepository.findByEmail(request.email)
            ?: throw BusinessException(ErrorCode.ADMIN_003)

        if (admin.status != AdminStatus.ACTIVE) {
            throw BusinessException(ErrorCode.ADMIN_001, mapOf(
                "message" to "비활성화된 관리자 계정입니다."
            ))
        }

        val isPasswordValid = CryptoUtil.verifyPassword(request.password, admin.passwordHash)
        if (!isPasswordValid) {
            throw BusinessException(ErrorCode.ADMIN_003)
        }

        // Update last login
        admin.lastLoginAt = LocalDateTime.now()
        adminUserRepository.save(admin)

        // Generate admin token
        val accessToken = jwtTokenProvider.generateAccessToken(admin.id.toString(), admin.email)

        return AdminLoginResponse(
            accessToken = accessToken,
            admin = AdminInfo(
                id = admin.id,
                email = admin.email,
                name = admin.name,
                role = admin.role
            )
        )
    }

    @Transactional(readOnly = true)
    fun getDashboardStats(): DashboardStats {
        val today = LocalDate.now().atStartOfDay()
        val firstDayOfMonth = LocalDate.now().withDayOfMonth(1).atStartOfDay()

        val totalUsers = userRepository.count().toInt()
        val activeUsers = userRepository.findAll().count { it.status == UserStatus.ACTIVE }
        val newUsersToday = userRepository.findAll().count { it.createdAt.isAfter(today) }
        val newUsersThisMonth = userRepository.findAll().count { it.createdAt.isAfter(firstDayOfMonth) }
        val totalFamilyGroups = familyGroupRepository.count().toInt()
        val totalMissions = missionRepository.count().toInt()
        val completedMissions = missionRepository.findAll().count { it.status == MissionStatus.COMPLETED }
        val totalBadgesIssued = userBadgeRepository.count().toInt()

        val missionCompletionRate = if (totalMissions > 0) {
            ((completedMissions.toDouble() / totalMissions) * 100).toInt()
        } else {
            0
        }

        return DashboardStats(
            totalUsers = totalUsers,
            activeUsers = activeUsers,
            newUsersToday = newUsersToday,
            newUsersThisMonth = newUsersThisMonth,
            totalFamilyGroups = totalFamilyGroups,
            totalMissions = totalMissions,
            completedMissions = completedMissions,
            missionCompletionRate = missionCompletionRate,
            totalBadgesIssued = totalBadgesIssued
        )
    }

    @Transactional(readOnly = true)
    fun getUsers(query: UserQueryRequest): List<User> {
        var users = userRepository.findAll()

        query.search?.let { search ->
            users = users.filter {
                it.email.contains(search, ignoreCase = true) ||
                it.name.contains(search, ignoreCase = true)
            }
        }

        query.status?.let { status ->
            users = users.filter { it.status == status }
        }

        return users.sortedByDescending { it.createdAt }
    }

    @Transactional
    fun updateUserStatus(userId: UUID, status: UserStatus): User {
        val user = userRepository.findById(userId)
            .orElseThrow { BusinessException(ErrorCode.USER_003) }
        user.status = status
        return userRepository.save(user)
    }

    @Transactional(readOnly = true)
    fun getCategories(): List<LifeCycleCategory> {
        return categoryRepository.findAll().sortedBy { it.displayOrder }
    }

    @Transactional
    fun createCategory(request: CreateCategoryRequest): LifeCycleCategory {
        val category = LifeCycleCategory().apply {
            this.name = request.name
            this.displayOrder = request.displayOrder ?: 0
            this.status = com.nestack.common.enum.CategoryStatus.ACTIVE
        }
        return categoryRepository.save(category)
    }

    @Transactional
    fun updateCategory(categoryId: UUID, request: UpdateCategoryRequest): LifeCycleCategory {
        val category = categoryRepository.findById(categoryId)
            .orElseThrow { BusinessException(ErrorCode.MISSION_002) }

        request.name?.let { category.name = it }
        request.displayOrder?.let { category.displayOrder = it }
        request.status?.let { category.status = it }

        return categoryRepository.save(category)
    }

    @Transactional
    fun deleteCategory(categoryId: UUID) {
        val category = categoryRepository.findById(categoryId)
            .orElseThrow { BusinessException(ErrorCode.MISSION_002) }
        categoryRepository.delete(category)
    }

    @Transactional(readOnly = true)
    fun getTemplates(): List<MissionTemplate> {
        return templateRepository.findAll()
    }

    @Transactional
    fun createTemplate(request: CreateTemplateRequest): MissionTemplate {
        val template = MissionTemplate().apply {
            this.name = request.name
            this.description = request.description
            this.categoryId = request.categoryId
            this.goalType = request.goalType ?: com.nestack.common.enum.GoalType.AMOUNT
            this.defaultGoalAmount = request.defaultGoalAmount
            this.status = com.nestack.common.enum.CategoryStatus.ACTIVE
        }
        return templateRepository.save(template)
    }

    @Transactional
    fun updateTemplate(templateId: UUID, request: UpdateTemplateRequest): MissionTemplate {
        val template = templateRepository.findById(templateId)
            .orElseThrow { BusinessException(ErrorCode.MISSION_002) }

        request.name?.let { template.name = it }
        request.description?.let { template.description = it }
        request.goalType?.let { template.goalType = it }
        request.defaultGoalAmount?.let { template.defaultGoalAmount = it }
        request.status?.let { template.status = it }

        return templateRepository.save(template)
    }

    @Transactional
    fun deleteTemplate(templateId: UUID) {
        val template = templateRepository.findById(templateId)
            .orElseThrow { BusinessException(ErrorCode.MISSION_002) }
        templateRepository.delete(template)
    }

    @Transactional(readOnly = true)
    fun getBadges(): List<Badge> {
        return badgeRepository.findAll()
    }

    @Transactional
    fun createBadge(request: CreateBadgeRequest): Badge {
        val badge = Badge().apply {
            this.name = request.name
            this.description = request.description
            this.imageUrl = request.imageUrl
            this.badgeType = request.badgeType
            this.conditionType = request.conditionType
            this.conditionValue = request.conditionValue
            this.status = com.nestack.common.enum.CategoryStatus.ACTIVE
        }
        return badgeRepository.save(badge)
    }

    @Transactional
    fun updateBadge(badgeId: UUID, request: UpdateBadgeRequest): Badge {
        val badge = badgeRepository.findById(badgeId)
            .orElseThrow { BusinessException(ErrorCode.BADGE_001) }

        request.name?.let { badge.name = it }
        request.description?.let { badge.description = it }
        request.imageUrl?.let { badge.imageUrl = it }
        request.conditionType?.let { badge.conditionType = it }
        request.conditionValue?.let { badge.conditionValue = it }
        request.status?.let { badge.status = it }

        return badgeRepository.save(badge)
    }

    @Transactional
    fun deleteBadge(badgeId: UUID) {
        val badge = badgeRepository.findById(badgeId)
            .orElseThrow { BusinessException(ErrorCode.BADGE_001) }
        badgeRepository.delete(badge)
    }

    @Transactional
    fun issueBadge(adminId: UUID, request: IssueBadgeRequest) {
        badgeService.awardBadge(
            userId = request.userId,
            badgeId = request.badgeId,
            issueType = com.nestack.common.enum.BadgeIssueType.MANUAL,
            issuedBy = adminId
        )
    }

    @Transactional(readOnly = true)
    fun getAnnouncements(): List<Announcement> {
        return announcementRepository.findAll().sortedByDescending { it.createdAt }
    }

    @Transactional
    fun createAnnouncement(adminId: UUID, request: CreateAnnouncementRequest): Announcement {
        val announcement = Announcement().apply {
            this.title = request.title
            this.content = request.content
            this.displayType = request.displayType
            this.startDate = request.startDate
            this.endDate = request.endDate
            this.status = com.nestack.common.enum.AnnouncementStatus.ACTIVE
            this.createdBy = adminId
        }
        return announcementRepository.save(announcement)
    }

    @Transactional
    fun updateAnnouncement(announcementId: UUID, request: UpdateAnnouncementRequest): Announcement {
        val announcement = announcementRepository.findById(announcementId)
            .orElseThrow { BusinessException(ErrorCode.COMMON_001) }

        request.title?.let { announcement.title = it }
        request.content?.let { announcement.content = it }
        request.displayType?.let { announcement.displayType = it }
        request.startDate?.let { announcement.startDate = it }
        request.endDate?.let { announcement.endDate = it }
        request.status?.let { announcement.status = it }

        return announcementRepository.save(announcement)
    }

    @Transactional
    fun deleteAnnouncement(announcementId: UUID) {
        val announcement = announcementRepository.findById(announcementId)
            .orElseThrow { BusinessException(ErrorCode.COMMON_001) }
        announcementRepository.delete(announcement)
    }

    @Transactional(readOnly = true)
    fun getActiveAnnouncements(): List<Announcement> {
        val now = LocalDateTime.now()
        return announcementRepository.findActiveAnnouncements(
            com.nestack.common.enum.AnnouncementStatus.ACTIVE,
            now
        )
    }

    data class DashboardStats(
        val totalUsers: Int,
        val activeUsers: Int,
        val newUsersToday: Int,
        val newUsersThisMonth: Int,
        val totalFamilyGroups: Int,
        val totalMissions: Int,
        val completedMissions: Int,
        val missionCompletionRate: Int,
        val totalBadgesIssued: Int
    )
}

File: nestack-backend/src/modules/admin/admin.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, MoreThanOrEqual, LessThanOrEqual, And } from 'typeorm';
import { JwtService } from '@nestjs/jwt';
import {
  AdminUser,
  User,
  FamilyGroup,
  Mission,
  LifeCycleCategory,
  MissionTemplate,
  Badge,
  UserBadge,
  Announcement,
} from '../../database/entities';
import {
  UserStatus,
  FamilyStatus,
  MissionStatus,
  CategoryStatus,
  BadgeIssueType,
  AnnouncementStatus,
} from '../../common/enums';
import {
  InvalidCredentialsException,
  UserNotFoundException,
  BadgeNotFoundException,
  CategoryNotFoundException,
  TemplateNotFoundException,
  AnnouncementNotFoundException,
} from '../../common/exceptions/business.exception';
import { comparePassword, hashPassword } from '../../common/utils/crypto.util';
import {
  AdminLoginDto,
  AdminTokenResponseDto,
  DashboardStatsDto,
  CreateCategoryDto,
  UpdateCategoryDto,
  CreateTemplateDto,
  UpdateTemplateDto,
  CreateBadgeDto,
  UpdateBadgeDto,
  IssueBadgeDto,
  CreateAnnouncementDto,
  UpdateAnnouncementDto,
} from './dto';

@Injectable()
export class AdminService {
  private readonly logger = new Logger(AdminService.name);

  constructor(
    @InjectRepository(AdminUser)
    private adminUserRepository: Repository<AdminUser>,
    @InjectRepository(User)
    private userRepository: Repository<User>,
    @InjectRepository(FamilyGroup)
    private familyGroupRepository: Repository<FamilyGroup>,
    @InjectRepository(Mission)
    private missionRepository: Repository<Mission>,
    @InjectRepository(LifeCycleCategory)
    private categoryRepository: Repository<LifeCycleCategory>,
    @InjectRepository(MissionTemplate)
    private templateRepository: Repository<MissionTemplate>,
    @InjectRepository(Badge)
    private badgeRepository: Repository<Badge>,
    @InjectRepository(UserBadge)
    private userBadgeRepository: Repository<UserBadge>,
    @InjectRepository(Announcement)
    private announcementRepository: Repository<Announcement>,
    private jwtService: JwtService,
  ) {}

  // Auth
  async login(dto: AdminLoginDto): Promise<AdminTokenResponseDto> {
    const admin = await this.adminUserRepository.findOne({
      where: { email: dto.email },
    });

    if (!admin) {
      throw new InvalidCredentialsException();
    }

    const isPasswordValid = await comparePassword(dto.password, admin.passwordHash);
    if (!isPasswordValid) {
      throw new InvalidCredentialsException();
    }

    // Update last login
    admin.lastLoginAt = new Date();
    await this.adminUserRepository.save(admin);

    const token = this.jwtService.sign({
      sub: admin.id,
      email: admin.email,
      role: admin.role,
      type: 'admin',
    });

    return {
      accessToken: token,
      tokenType: 'Bearer',
      expiresIn: 86400, // 24 hours
      admin: {
        id: admin.id,
        email: admin.email,
        name: admin.name,
        role: admin.role,
      },
    };
  }

  // Dashboard
  async getDashboardStats(): Promise<DashboardStatsDto> {
    const today = new Date();
    today.setHours(0, 0, 0, 0);

    const [
      totalUsers,
      activeUsers,
      newUsersToday,
      totalFamilyGroups,
      totalMissions,
      completedMissions,
      activeBadges,
    ] = await Promise.all([
      this.userRepository.count(),
      this.userRepository.count({ where: { status: UserStatus.ACTIVE } }),
      this.userRepository.count({
        where: { createdAt: MoreThanOrEqual(today) },
      }),
      this.familyGroupRepository.count({
        where: { status: FamilyStatus.ACTIVE },
      }),
      this.missionRepository.count(),
      this.missionRepository.count({
        where: { status: MissionStatus.COMPLETED },
      }),
      this.badgeRepository.count({ where: { status: CategoryStatus.ACTIVE } }),
    ]);

    // Calculate total savings
    const missions = await this.missionRepository.find();
    const totalSavingsAmount = missions.reduce(
      (sum, m) => sum + Number(m.currentAmount),
      0,
    );

    return {
      totalUsers,
      activeUsers,
      newUsersToday,
      totalFamilyGroups,
      totalMissions,
      completedMissions,
      totalSavingsAmount,
      activeBadges,
    };
  }

  // Users
  async getUsers(page = 1, limit = 20): Promise<{ users: User[]; total: number }> {
    const [users, total] = await this.userRepository.findAndCount({
      order: { createdAt: 'DESC' },
      skip: (page - 1) * limit,
      take: limit,
    });

    return { users, total };
  }

  async updateUserStatus(userId: string, status: UserStatus): Promise<User> {
    const user = await this.userRepository.findOne({ where: { id: userId } });
    if (!user) {
      throw new UserNotFoundException();
    }

    user.status = status;
    return this.userRepository.save(user);
  }

  // Categories
  async getCategories(): Promise<LifeCycleCategory[]> {
    return this.categoryRepository.find({ order: { displayOrder: 'ASC' } });
  }

  async createCategory(dto: CreateCategoryDto): Promise<LifeCycleCategory> {
    const category = this.categoryRepository.create({
      name: dto.name,
      displayOrder: dto.displayOrder || 0,
      status: CategoryStatus.ACTIVE,
    });
    return this.categoryRepository.save(category);
  }

  async updateCategory(
    id: string,
    dto: UpdateCategoryDto,
  ): Promise<LifeCycleCategory> {
    const category = await this.categoryRepository.findOne({ where: { id } });
    if (!category) {
      throw new CategoryNotFoundException();
    }

    if (dto.name !== undefined) category.name = dto.name;
    if (dto.displayOrder !== undefined) category.displayOrder = dto.displayOrder;
    if (dto.status !== undefined) category.status = dto.status;

    return this.categoryRepository.save(category);
  }

  async deleteCategory(id: string): Promise<void> {
    const result = await this.categoryRepository.delete({ id });
    if (result.affected === 0) {
      throw new CategoryNotFoundException();
    }
  }

  // Templates
  async getTemplates(): Promise<MissionTemplate[]> {
    return this.templateRepository.find({
      relations: ['category'],
      order: { usageCount: 'DESC' },
    });
  }

  async createTemplate(dto: CreateTemplateDto): Promise<MissionTemplate> {
    const template = this.templateRepository.create({
      name: dto.name,
      description: dto.description,
      categoryId: dto.categoryId,
      goalType: dto.goalType,
      defaultGoalAmount: dto.defaultGoalAmount,
      status: CategoryStatus.ACTIVE,
    });
    return this.templateRepository.save(template);
  }

  async updateTemplate(
    id: string,
    dto: UpdateTemplateDto,
  ): Promise<MissionTemplate> {
    const template = await this.templateRepository.findOne({ where: { id } });
    if (!template) {
      throw new TemplateNotFoundException();
    }

    if (dto.name !== undefined) template.name = dto.name;
    if (dto.description !== undefined) template.description = dto.description;
    if (dto.categoryId !== undefined) template.categoryId = dto.categoryId;
    if (dto.goalType !== undefined) template.goalType = dto.goalType;
    if (dto.defaultGoalAmount !== undefined)
      template.defaultGoalAmount = dto.defaultGoalAmount;
    if (dto.status !== undefined) template.status = dto.status;

    return this.templateRepository.save(template);
  }

  async deleteTemplate(id: string): Promise<void> {
    const result = await this.templateRepository.delete({ id });
    if (result.affected === 0) {
      throw new TemplateNotFoundException();
    }
  }

  // Badges
  async getBadges(): Promise<Badge[]> {
    return this.badgeRepository.find({ order: { badgeType: 'ASC', name: 'ASC' } });
  }

  async createBadge(dto: CreateBadgeDto): Promise<Badge> {
    const badge = this.badgeRepository.create({
      name: dto.name,
      description: dto.description,
      imageUrl: dto.imageUrl,
      badgeType: dto.badgeType,
      conditionType: dto.conditionType,
      conditionValue: dto.conditionValue,
      status: CategoryStatus.ACTIVE,
    });
    return this.badgeRepository.save(badge);
  }

  async updateBadge(id: string, dto: UpdateBadgeDto): Promise<Badge> {
    const badge = await this.badgeRepository.findOne({ where: { id } });
    if (!badge) {
      throw new BadgeNotFoundException();
    }

    if (dto.name !== undefined) badge.name = dto.name;
    if (dto.description !== undefined) badge.description = dto.description;
    if (dto.imageUrl !== undefined) badge.imageUrl = dto.imageUrl;
    if (dto.badgeType !== undefined) badge.badgeType = dto.badgeType;
    if (dto.conditionType !== undefined) badge.conditionType = dto.conditionType;
    if (dto.conditionValue !== undefined) badge.conditionValue = dto.conditionValue;
    if (dto.status !== undefined) badge.status = dto.status;

    return this.badgeRepository.save(badge);
  }

  async deleteBadge(id: string): Promise<void> {
    await this.userBadgeRepository.delete({ badgeId: id });
    await this.badgeRepository.delete({ id });
  }

  async issueBadge(dto: IssueBadgeDto, issuedBy: string): Promise<UserBadge> {
    const user = await this.userRepository.findOne({
      where: { id: dto.userId },
    });
    if (!user) {
      throw new UserNotFoundException();
    }

    const badge = await this.badgeRepository.findOne({
      where: { id: dto.badgeId },
    });
    if (!badge) {
      throw new BadgeNotFoundException();
    }

    const userBadge = this.userBadgeRepository.create({
      userId: dto.userId,
      badgeId: dto.badgeId,
      issueType: BadgeIssueType.MANUAL,
      issuedBy,
    });

    return this.userBadgeRepository.save(userBadge);
  }

  // Announcements
  async getAnnouncements(): Promise<Announcement[]> {
    return this.announcementRepository.find({
      order: { startDate: 'DESC' },
      relations: ['createdByAdmin'],
    });
  }

  async getActiveAnnouncements(): Promise<Announcement[]> {
    const now = new Date();
    return this.announcementRepository.find({
      where: {
        status: AnnouncementStatus.ACTIVE,
        startDate: LessThanOrEqual(now),
        endDate: MoreThanOrEqual(now),
      },
      order: { startDate: 'DESC' },
    });
  }

  async createAnnouncement(
    dto: CreateAnnouncementDto,
    createdBy: string,
  ): Promise<Announcement> {
    const announcement = this.announcementRepository.create({
      title: dto.title,
      content: dto.content,
      displayType: dto.displayType,
      startDate: new Date(dto.startDate),
      endDate: new Date(dto.endDate),
      status: AnnouncementStatus.ACTIVE,
      createdBy,
    });
    return this.announcementRepository.save(announcement);
  }

  async updateAnnouncement(
    id: string,
    dto: UpdateAnnouncementDto,
  ): Promise<Announcement> {
    const announcement = await this.announcementRepository.findOne({
      where: { id },
    });
    if (!announcement) {
      throw new AnnouncementNotFoundException();
    }

    if (dto.title !== undefined) announcement.title = dto.title;
    if (dto.content !== undefined) announcement.content = dto.content;
    if (dto.displayType !== undefined) announcement.displayType = dto.displayType;
    if (dto.startDate !== undefined)
      announcement.startDate = new Date(dto.startDate);
    if (dto.endDate !== undefined) announcement.endDate = new Date(dto.endDate);
    if (dto.status !== undefined) announcement.status = dto.status;

    return this.announcementRepository.save(announcement);
  }

  async deleteAnnouncement(id: string): Promise<void> {
    const result = await this.announcementRepository.delete({ id });
    if (result.affected === 0) {
      throw new AnnouncementNotFoundException();
    }
  }
}
